services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/controller/SegmentController.java

```diff
@@ -0,0 +1,208 @@
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(segmentid);
+            SegmentWebRequestObject segmentWebRequestObject = resource.getSegment();
+            BeanUtils.copyProperties(segmentWebRequestObject, segmentWebResponseObject);
+            RestPreconditionsUtil.verifyResourceNotNull(segmentWebResponseObject);
+            RestPreconditionsUtil.populateResourceProjectId(segmentWebResponseObject, projectid);
+            RestPreconditionsUtil.populateResourceSegmentId(segmentWebResponseObject, segmentid);
+
+            segmentWebResponseObject = this.segmentDatabaseService.getBySegmentId(segmentid);
+            if (segmentWebResponseObject == null) {
+                throw new ResourceNotFoundException("Segment not found : " + segmentid);
+            }
+
+            this.segmentDatabaseService.addSegment(segmentWebResponseObject);
+
+            segmentWebResponseObject = this.segmentDatabaseService.getBySegmentId(segmentid);
+
+        } catch (ParameterNullOrEmptyException e) {
+            throw new Exception(e);
+        }
+
+        return new SegmentWebResponseJson(segmentWebResponseObject);
+
+    }
+
+    @RequestMapping(
+            method = DELETE,
+            value = {"/project/{projectid}/segments/{segmentid}", "/v4/{projectid}/segments/{segmentid}"})
+    public ResponseId deleteSegmentBySegmentId(@PathVariable String projectid, @PathVariable String segmentid) throws Exception {
+
+        SegmentWebResponseObject segmentWebResponseObject = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(segmentid);
+            RestPreconditionsUtil.verifyResourceFound(projectid);
+
+            segmentWebResponseObject = this.segmentDatabaseService.getBySegmentId(segmentid);
+            if (segmentWebResponseObject == null) {
+                return new ResponseId();
+            }
+
+            // Release Network Type
+            String networkType = segmentWebResponseObject.getNetworkType();
+            Long key = Long.parseLong(String.valueOf(segmentWebResponseObject.getSegmentationId()));
+            if (NetworkTypeEnum.VXLAN.getNetworkType().equals(networkType)) {
+                this.segmentService.releaseVxlanEntity(segmentWebResponseObject.getSegmentationUUID(), key);
+            } else if (NetworkTypeEnum.VLAN.getNetworkType().equals(networkType)) {
+                this.segmentService.releaseVlanEntity(segmentWebResponseObject.getSegmentationUUID(), key);
+            }else if (NetworkTypeEnum.GRE.getNetworkType().equals(networkType)) {
+                this.segmentService.releaseGreEntity(segmentWebResponseObject.getSegmentationUUID(), key);
+            }
+
+            segmentDatabaseService.deleteSegment(segmentid);
```

Comment:
Usually DELETE is the reverse order of POST. Let us do this:
(1) segmentDatabaseService.deleteSegment(segmentid);
(2) segmentService.release VxLan/VLan/Gre entity.

If segmentService tries to release entity first and the process crashes before segmentDatabaseService.deleteSegment, the segmentation Id might be released already so will be allocated to other segment, which causes "one key shared by two segments" issue.

Reply:
We might need to visit other Controller in VPC/subnet/Route to see if we hit the same issue elsewhere.